File: Data Structures & Algorithms/climbing-stairs/submission-1.py
class Solution:
        
    def solve(self,n:int, memo):
        if (n<=2):
            return n
        if n in memo:
            return memo[n]
        else:
            memo[n]=self.solve(n-1,memo)+self.solve(n-2,memo)
        return memo[n]


    def climbStairs(self, n: int) -> int:
        memo={}
        return self.solve(n,memo)


        

File: Data Structures & Algorithms/partition-equal-subset-sum/submission-1.cpp
class Solution {
public:
    bool solve(const vector<int>& nums, int target) {
        vector<bool> dp(target + 1, false);
        dp[0] = true;

        for (int num : nums) {
            for (int j = target; j >= num; j--) {
                dp[j] = dp[j] || dp[j - num];
            }
        }

        return dp[target];
    }

    bool canPartition(vector<int>& nums) {
        int sum = 0;

        for (int i = 0; i < nums.size(); i++) {
            sum += nums[i];
        }

        if (sum % 2 != 0) {
            return false;
        }

        return solve(nums, sum / 2);
    }
};